chips/stm32f429zi/src/ethernet/ethernet.rs

```diff
@@ -0,0 +1,2288 @@
+        self.dma_registers.dmasr.is_set(DMASR::TBUS)
+    }
+
+    fn clear_transmit_buffer_unavailable_interrupt(&self) {
+        self.dma_registers.dmasr.modify(DMASR::TBUS::SET);
+    }
+
+    fn did_transmit_process_stopped_interrupt_occur(&self) -> bool {
+        self.dma_registers.dmasr.is_set(DMASR::TPSS)
+    }
+
+    fn clear_transmit_process_stopped_interrupt(&self) {
+        self.dma_registers.dmasr.modify(DMASR::TPSS::SET);
+    }
+
+    fn did_transmit_interrupt_occur(&self) -> bool {
+        self.dma_registers.dmasr.is_set(DMASR::TS)
+    }
+
+    fn clear_transmit_interrupt(&self) {
+        self.dma_registers.dmasr.modify(DMASR::TS::SET);
+    }
+
+    fn enable_transmit_store_and_forward(&self) -> Result<(), ErrorCode> {
+        if self.get_transmit_process_state() != DmaTransmitProcessState::Stopped {
+            return Err(ErrorCode::FAIL);
+        }
+
+        self.dma_registers.dmaomr.modify(DMAOMR::TSF::SET);
+
+        Ok(())
+    }
+
+    fn disable_transmit_store_and_forward(&self) -> Result<(), ErrorCode> {
+        if self.get_transmit_process_state() != DmaTransmitProcessState::Stopped {
+            return Err(ErrorCode::FAIL);
+        }
+
+        self.dma_registers.dmaomr.modify(DMAOMR::TSF::CLEAR);
+
+        Ok(())
+    }
+
+    fn is_transmit_store_and_forward_enabled(&self) -> bool {
+        self.dma_registers.dmaomr.is_set(DMAOMR::TSF)
+    }
+
+    fn disable_flushing_received_frames(&self) {
+        self.dma_registers.dmaomr.modify(DMAOMR::DFRF::SET);
+    }
+
+    fn forward_error_frames(&self) {
+        self.dma_registers.dmaomr.modify(DMAOMR::FEF::SET);
+    }
+
+    fn forward_undersized_good_frames(&self) {
+        self.dma_registers.dmaomr.modify(DMAOMR::FUGF::SET);
+    }
+
+    fn enable_receive_store_and_forward(&self) -> Result<(), ErrorCode> {
+        if self.get_receive_process_state() != DmaReceiveProcessState::Stopped {
+            return Err(ErrorCode::FAIL);
+        }
+
+        self.dma_registers.dmaomr.modify(DMAOMR::RSF::SET);
+
+        Ok(())
+    }
+
+    fn disable_receive_store_and_forward(&self) -> Result<(), ErrorCode> {
+        if self.get_receive_process_state() != DmaReceiveProcessState::Stopped {
+            return Err(ErrorCode::FAIL);
+        }
+
+        self.dma_registers.dmaomr.modify(DMAOMR::RSF::CLEAR);
+
+        Ok(())
+    }
+
+    fn is_receive_store_and_forward_enabled(&self) -> bool {
+        self.dma_registers.dmaomr.is_set(DMAOMR::RSF)
+    }
+
+    fn flush_dma_transmit_fifo(&self) -> Result<(), ErrorCode> {
+        if self.get_transmit_process_state() != DmaTransmitProcessState::Stopped {
+            return Err(ErrorCode::FAIL);
+        }
+
+        self.dma_registers.dmaomr.modify(DMAOMR::FTF::SET);
+
+        // TODO: Adjust this value
+        for _ in 0..100 {
+            if self.dma_registers.dmaomr.read(DMAOMR::FTF) == 0 {
+                return Ok(());
+            }
+        }
+
+        Err(ErrorCode::BUSY)
```

Comment:
For such methods featuring a timeout, it'd be great to document the exact reasons why this could fail, what the magic timeout parameter means (i.e. why you came up with this value, even if it just happens to work in practice that's worth documenting), and perhaps make it configurable via a parameter.

Reply:
Alternatively, if this is really guaranteed to succeed _eventually_ it's OK to loop until the condition is true. E.g., some of the GPIO hardware (such as the SAM4L) have a busy bit which will eventually clear, so the drivers for that hardware just loops on the busy bit before reading or setting the GPIO state and there is no error condition.

Reply:
Some of these for loops can fail if the peripheral mode configuration register doesn't match the actual hardware. Not sure if I should replace them with while loops since by the time the MAC and DMA cores are initialized, `syscfg` is already configured in the right mode. The sole benefit of keeping a for loop would be spotting easier the case where the syscfg is not configured correctly, since this DMA/MAC core seems to be present on multiple boards Tock supports (MSP432 for instance).